packages/pros/src/io/print_impl.rs

```diff
@@ -0,0 +1,318 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![no_std]
+#![allow(dead_code)]
+#![allow(unused)]
+#![warn(unsafe_op_in_unsafe_fn)]
+
+use core::{convert::TryFrom, file, line, stringify};
+
+/// This forces a "C" library linkage
+#[cfg(not(windows))]
+#[link(name = "c")]
+mod c {
+    extern "C" {}
+}
+
+// These constants are used by the macros but we don't want to expose
+// them to library users.
+#[doc(hidden)]
+pub const __NEWLINE: &str = "\n";
+#[doc(hidden)]
+pub const __STDOUT: i32 = 1;
+#[doc(hidden)]
+pub const __STDERR: i32 = 2;
+
+#[doc(hidden)]
+pub struct __ProsSysWriter(i32);
+
+impl core::fmt::Write for __ProsSysWriter {
+    #[inline]
+    fn write_str(&mut self, s: &str) -> core::fmt::Result {
+        __println(self.0, s)
+    }
+}
+
+impl __ProsSysWriter {
+    #[inline]
+    pub fn new(handle: i32) -> __ProsSysWriter {
+        __ProsSysWriter(handle)
+    }
+
+    #[inline]
+    pub fn write_fmt(&mut self, args: core::fmt::Arguments) -> core::fmt::Result {
+        core::fmt::Write::write_fmt(self, args)
+    }
+
+    #[inline]
+    pub fn write_str(&mut self, s: &str) -> core::fmt::Result {
+        __println(self.0, s)
+    }
+
+    #[inline]
+    pub fn write_nl(&mut self) -> core::fmt::Result {
+        __println(self.0, __NEWLINE)
+    }
+}
+
+#[doc(hidden)]
+#[inline]
+pub fn __println(handle: i32, msg: &str) -> core::fmt::Result {
+    let msg = msg.as_bytes();
+
+    let mut written = 0;
+    while written < msg.len() {
+        match unsafe { write(handle, &msg[written..]) } {
+            // Ignore errors
+            None | Some(0) => break,
+            Some(res) => written += res,
+        }
+    }
+
+    Ok(())
+}
+
+unsafe fn write(handle: i32, bytes: &[u8]) -> Option<usize> {
+    usize::try_from(unsafe {
+        pros_sys::write(
+            handle,
+            bytes.as_ptr().cast::<core::ffi::c_void>(),
+            bytes.len(),
+        )
+    })
+    .ok()
+}
+
+/// Macro for printing to the standard output, with a newline.
+///
+/// Does not panic on failure to write - instead silently ignores errors.
+///
+/// See [`println!`](https://doc.rust-lang.org/std/macro.println.html) for
+/// full documentation.
```

Comment:
Does everything in the standard library docs actually apply to these macros?

Reply:
Presumably, since its using the same formatting API under the hood.